cpu/esp32,boards/*esp32*:  fix documentation (#16952)
RIOT-OS/RIOT: cpu/esp32/doc.txt

    4. [ESP-NOW Network Interface](#esp32_esp_now_network_interface)
    5. [Other Network Devices](#esp32_other_network_devices)
10. [Application-Specific Configurations](#esp32_application_specific_configurations)
    1. [Make Variable `CONFIGS`](#esp32_config_make_variable)
    2. [Application-Specific Board Configuration](#esp32_application_specific_board_configuration)
    3. [Application-Specific Driver Configuration](#esp32_application_specific_driver_configuration)
11. [Debugging](#esp32_debugging)
    1. [JTAG Debugging](#esp32_jtag_debugging)
    2. [QEMU Mode and GDB](#esp32_qemu_mode_and_gdb)

> # Overview {#esp32_overview}

Reviewer: In the previous documentation, each section title had a link to the table of contents. This was very helpful because the documentation is quite large and the user was able jump back to the table of contents from anywhere in the large document using these links. Would it be possible to add these links again?

Author: Maybe just adding
```
[Back to table of contents](#esp32_toc)
```
at the bottom of every (sub-)section just before the next section header would be enough? Syntaxwise, that should be pretty future safe regarding updates of Doxygen.

| IR                | up to 8 channels TX/RX                                            | no |
| Motor PWM         | 2 devices x 6 channels                                            | yes |
| LED PWM           | 16 channels                                                       | no |
| Crypto            | Hardware acceleration of AES, SHA-2, RSA, ECC, RNG                | no |
| Vcc               | 2.5 - 3.6 V                                                       | |
| Documents         | [Datasheet](https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf) <br> [Technical Reference](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf) | |

</center><br>

@note Even if used ESP32 SoC is a dual-core version, RIOT-OS uses only one core.

Rather than using the ESP32 SoC directly, ESP32 boards use an
[ESP32 module from Espressif](https://www.espressif.com/en/products/hardware/modules) which
integrates additionally to the SoC some key components, like SPI flash memory, SPI RAM, or crystal
oscillator. Some of these components are optional. A good overview about available modules can be

> found in the [Online documentation of ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/latest/hw-reference/modules-and-boards.html#wroom-solo-and-wrover-modules).

Reviewer: ```suggestion
found in the [Online documentation of ESP-IDF](https://www.espressif.com/en/products/modules).
```

Most common modules used by ESP32 boards are the [ESP32-WROOM-32](https://docs.espressif.com/projects/esp-idf/en/latest/hw-reference/modules-and-boards.html#esp32-wroom-32) and [ESP32-WROVER](https://docs.espressif.com/projects/esp-idf/en/latest/hw-reference/modules-and-boards.html#esp32-wrover).

> 

Reviewer: The references should be fixed:
```diff
- [ESP32-WROOM-32](https://docs.espressif.com/projects/esp-idf/en/latest/hw-reference/modules-and-boards.html#esp32-wroom-32)
+ [ESP32-WROOM-32](https://www.espressif.com/sites/default/files/documentation/esp32-wroom-32_datasheet_en.pdf)
```
```diff
- [ESP32-WROVER](https://docs.espressif.com/projects/esp-idf/en/latest/hw-reference/modules-and-boards.html#esp32-wrover)
+ [ESP32-WROVER](http://espressif.com/sites/default/files/documentation/esp32-wrover_datasheet_en.pdf)
```

additional information: the type of message, the system time in ms, and the
module or function in which the log message is generated. For example:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
I (663) [main_trampoline] main(): This is RIOT! (Version: 2019.10-devel-437-gf506a)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Either the `LOG_*` macros as defined in

> [system logging header](http://doc.riot-os.org/log_8h.html) or the tagged

Reviewer: ```diff
- [system logging header](http://doc.riot-os.org/log_8h.html)
+ [system logging header](https://doc.riot-os.org/core_2include_2log_8h.html)
```

Author: I added an anchor to the doc there and reference this instead, as future Doxygen versions might produce slightly different links. The anchor should hopefully remain stable.

infrastructure WiFi network. All ESP-NOW nodes must therefore be compiled with
the channel of the AP asvalue for the parameter 'ESP_NOW_CHANNEL'.

## Other Network Devices {#esp32_other_network_devices}

RIOT provides a number of driver modules for different types of network
devices, e.g., IEEE 802.15.4 radio modules and Ethernet modules. The RIOT port
for ESP32 has been tested with the following network devices:

> - [mrf24j40](https://riot-os.org/api/group__drivers__mrf24j40.html) (driver for

Author: Again, I reference the driver Doxygen group instead

  Microchip MRF24j40 based IEEE 802.15.4
- [enc28j60](https://riot-os.org/api/group__drivers__enc28j60.html) (driver for
  Microchip ENC28J60 based Ethernet modules)

> ### Using MRF24J40 (module `mrf24j40`) {#esp32_using_mrf24j40}

Reviewer: ```suggestion
### Using MRF24J40 (module mrf24j40) {#esp32_using_mrf24j40}
```

To use MRF24J40 based IEEE 802.15.4 modules as network device, the
`mrf24j40` driver module has to be added to the makefile of the
application:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
USEMODULE += mrf24j40
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The driver parameters that have to be defined by the board configuration for
the MRF24J40 driver module are:

Parameter              | Description
-----------------------|------------
MRF24J40_PARAM_CS      | GPIO used as CS signal
MRF24J40_PARAM_INT     | GPIO used as interrupt signal
MRF24J40_PARAM_RESET   | GPIO used as reset signal

Since each board has different GPIO configurations, refer to the board
documentation for the GPIOs recommended for the MRF24J40.

@note The reset signal of the MRF24J40 based network device can be connected
with the ESP32 RESET/EN pin  which is broken out on most boards. This keeps the
GPIO free defined by `MRF24J40_PARAM_RESET` free for other purposes.

> ### Using ENC28J60 (module `enc28j60`) {#esp32_using_enc28j60}

Reviewer: ```suggestion
### Using ENC28J60 (module enc28j60) {#esp32_using_enc28j60}
```

with the ESP32 RESET/EN pin  which is broken out on most boards. This keeps the
GPIO free defined by `ENC28J60_PARAM_RESET` free for other purposes.

# Application-Specific Configurations {#esp32_application_specific_configurations}

The board-specific configuration files `board.h` and `periph_conf.h` as
well well as the driver parameter configuration files `<driver>_params.h`
define the default configurations for peripherals and device driver modules.
These are, for example, the GPIOs used, bus interfaces used or available bus
speeds. Because there are many possible configurations and many different
application requirements, these default configurations are usually only a
compromise between different requirements.

Therefore, it is often necessary to change some of these default configurations
for individual applications. For example, while many PWM channels are needed in
one application, another application does not need PWM channels, but many ADC
channels.

There are two ways to give the application the ability to change some of these
default configurations:

- make variable `CFLAGS`
- application-specific board or driver configuration file

> ## Make Variable `CFLAGS` {#esp32_config_make_variable}

Reviewer: ```suggestion
## Make Variable CFLAGS {#esp32_config_make_variable}
```
